ee/tabby-ui/components/chat/use-tabby-answer.ts

```diff
@@ -0,0 +1,174 @@
+import React from 'react'
+import { mergeWith } from 'lodash-es'
+
+import { AnswerRequest, AnswerResponse } from '@/lib/types'
+
+const NEWLINE = '\n'.charCodeAt(0)
+
+interface UseTabbyAnswerOptions {
+  onError?: (err: Error) => void
+  api?: string
+  fetcher?: typeof fetch
+  headers?: Record<string, string> | Headers
+}
+
+export function useTabbyAnswer({
+  api = '/v1beta/answer',
+  onError,
+  headers,
+  fetcher
+}: UseTabbyAnswerOptions) {
+  const [isLoading, setIsLoading] = React.useState(false)
+  const [answer, setAnswer] = React.useState<AnswerResponse | undefined>()
+  const [error, setError] = React.useState<Error | undefined>()
+  // Abort controller to cancel the current API call.
+  const abortControllerRef = React.useRef<AbortController | null>(null)
+
+  const stop = () => {
+    if (abortControllerRef.current) {
+      abortControllerRef.current.abort()
+      abortControllerRef.current = null
+    }
+  }
+
+  const mergeParsedData = (
+    data1: Partial<AnswerResponse>,
+    data2: Partial<AnswerResponse>
+  ): Partial<AnswerResponse> => {
+    return mergeWith({}, data1, data2, (obj, src) => {
+      if (typeof obj === 'string' && typeof src === 'string') {
+        return obj + src
+      }
+    })
+  }
+
+  const triggerRequest = async (data: AnswerRequest) => {
+    try {
+      setIsLoading(true)
+      setError(undefined)
+      setAnswer(undefined)
+      const abortController = new AbortController()
+      abortControllerRef.current = abortController
+
+      const fetch = fetcher ?? window.fetch
+
+      const response = await fetch(api, {
+        method: 'POST',
+        body: JSON.stringify(data),
+        headers: {
+          'Content-Type': 'application/json',
+          ...headers
+        },
+        signal: abortController?.signal
+      }).catch(err => {
+        throw err
+      })
+
+      if (!response?.ok) {
+        throw new Error(String(response.status))
+      }
+
+      if (response.body == null) {
+        throw new Error('The response body is empty')
+      }
+
+      for await (const line of readDataStream(response.body.getReader(), {
```

Comment:
consider using a event stream parser, @icycodes happen to investigate this area recently

Reply:
using 'eventsource-parser'